src/db/gc/mod.rs

```diff
@@ -0,0 +1,323 @@
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+//!
+//! The current implementation of the garbage collector is `concurrent mark-and-sweep`.
+//!
+//! ## Terminology
+//! `chain finality` - a number of epochs after which it becomes impossible to add or remove a block
+//! previously appended to the blockchain.
+//!
+//! ## Design goals
+//! A correct GC algorithm that is simple and efficient for forest scenarios. This algorithm removes
+//! unreachable blocks that are older than `chain finality`, making sure to avoid removing something
+//! that could later become reachable as a result of a fork.
+//!
+//! ## GC Algorithm
+//! The `mark-and-sweep` algorithm was chosen due to it's simplicity, efficiency and low memory
+//! footprint. Previously the `semi-space` algorithm was used resulting in data duplication and up
+//! to a 100% extra disk usage.
+//!
+//! ## GC Workflow
+//! 1. Mark: traverse all the relevant database columns, generating integer hash representations for
+//! each database key and storing those in a set.
+//! 2. Wait at least `chain finality` blocks.
+//! 3. Traverse reachable blocks starting at the current heaviest tipset and remove those from the
+//! marked set, leaving only unreachable entries that are older than `chain finality`.
+//! 4. Sweep, removing all the remaining marked entries from the database.
+//!
+//! ## Correctness
+//! This algorithm considers all the blocks that are visited during the `snapshot export` task
+//! reachable, making sure they are kept in the database after the run. It makes sure to retain the
+//! reachable graph as well as all the blocks for at least `chain finality` to account for potential
+//! forks. A snapshot can be used to bootstrap the node from scratch, thus the algorithm is
+//! considered correct when a valid snapshot can be exported using records available in the database
+//! after the run.
+//!
+//! ## Disk usage
+//! There's no additional disk space required to run this algorithm. However, removing the
+//! unreachable blocks from the database takes at least `chain finality`. The GC speed depends on
+//! the reachable graph size. Disk usage is also affected by the GC run interval. Additionally,
+//! since a truncated 4-byte hash is used - there's a slight possibility of a collision, which might
+//! result in an unreachable block being retained in the database. Still, the impact on the total
+//! disk size is negligible.
+//!
+//! ## Memory usage
+//! During the `mark` and up to the `sweep` stage, the algorithm requires `4 bytes` of memory for
+//! each database record. Additionally, the seen cache while traversing the reachable graph
+//! executing the `filter` stage requires at least `32 bytes` of memory for each reachable block.
```

Comment:
Let's put some real-world numbers on this. What is the expected memory usage with a 100GiB database (roughly the size of a mainnet snapshot in ParityDB)?

Reply:
This actually needs to be measured.

Reply:
Done

Reply:
Great!